### PR TITLE
Adding absorption and xp API, and weather control

### LINKED ISSUE #537: Suggestion: API function to query and modify entity's absorption level
Self-explanatory. `query(e, 'absorption')` and `modify(e, 'absorption', number)` should work basically the same way as the corresponding functions as health, except they act on the absorption health of an entity (yellow hearts above health bar in HUD).

### LINKED ISSUE #539: Suggestion: API function to query and modify player's XP stats
Something like:
`query(e, 'xp_level')`: Gets the player's XP level.
`query(e, 'xp_points')`: Gets the player's progress to next XP level (as a fractional number from 0-1)
`query(e, 'score')`: Gets the player's score (a number that's displayed on the death screen).
`modify(e, 'xp_level', number)`: Sets the player's XP level.
`modify(e, 'xp_points', number)`: Sets the player's progress to next XP level (as a fractional number from 0-1)
`modify(e, 'score', number)`: Sets the player's score.

An example use case is to provide an alternative XP system different from the standard XP level formula.

### LINKED ISSUE #547: Scarpet: Missing support for wether?
Both querying and manipulating wether through scarpet would be useful. Support for interacting with the gamerule(s) could be useful too, besides jsut using `run('gamerule ...')`.

### PR DESCRIPTION
**Pull request by Ghoulboy78**
Fixes #537 , fixes #539, fixes #547  and also adds ability to print to the action bar.

### CODE REVIEW

**Comment by altrisi** [comment]
Psst! For your next PR, there's no longer the need to update `Full.md` because it's now automatically merged via Github Actions on merge (since d8009a37196e244c29b10b78eb53b92573cd1b3a).

That simplifies both your work and the work of the reviewer (no need for u to change it, no need to review it's correct and less clutter in the PR).

###### (u don't need to remove it now, just note that if it finds that it changed it won't run)

**Comment by Ghoulboy78** [comment]
Yo that's great man.

File: src/main/java/carpet/script/value/EntityValue.java [hunk 1]
```diff
@@ -476,6 +476,31 @@ public Value get(String what, Value arg)
             return Value.NULL;
         });
 
+        put("absorption",(e, a)->{
+            if(e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).getAbsorptionAmount());
+            return Value.NULL;
+        });
```
Comment (James103) [inline src/main/java/carpet/script/value/EntityValue.java:482]: There's no need to restrict setting absorption level to players only since all living entities have the capability to have absorption points.

File: src/main/java/carpet/script/value/EntityValue.java [hunk 2]
```diff
@@ -1294,6 +1319,30 @@ else if (v instanceof ListValue)
             if(e instanceof PlayerEntity) ((PlayerEntity) e).getHungerManager().addExhaustion((int) NumericValue.asNumber(v).getLong());
         });
 
+        put("absorption",(e, v)->{
+            if(e instanceof PlayerEntity) ((PlayerEntity) e).setAbsorptionAmount((float) NumericValue.asNumber(v).getLong());
+        });
+
```
Comment (James103) [inline src/main/java/carpet/script/value/EntityValue.java:1325]: There's no need to restrict setting absorption level to players only since all living entities have the capability to have absorption points.

File: src/main/java/carpet/script/value/EntityValue.java (hunk 1 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:482]: huh I didnt know that.

Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:482]: Umm, no. That method only exists in PlayerEntity.java, not Entity.java. Entities don't have absorption.

**Comment by Ghoulboy78** [comment]
Umm, @altrisi , it ain't updating Full.md...

**Comment by altrisi** [comment]
The action only runs when merging so the changes don't clutter the `Files changed` tab (and it doesn't run all the time).

However, I think this time it won't run anyway since `Full.md` is already modified (it has a filter to not run if it also changed, not sure why did I add that).

File: src/main/java/carpet/script/value/EntityValue.java [hunk 3]
```diff
@@ -1294,6 +1319,30 @@ else if (v instanceof ListValue)
...
+        put("add_score",(e, v)->{
+            if(e instanceof PlayerEntity) ((PlayerEntity) e).addScore((int) NumericValue.asNumber(v).getLong());
+        });
+        /*
+        put("xp_points",(e, v)->{//idk how useful tbh, and dont work anyways.
+            if(e instanceof PlayerEntity((PlayerEntity) e).experienceProgress=(int) NumericValue.asNumber(v).getLong();
+        });
+        */
```
Comment (altrisi) [inline src/main/java/carpet/script/value/EntityValue.java:1345]: This doesn't work (as far as I understood) because `experienceProgress` is only for getting the progress bar to be nice, but it is reset whenever some change to XP happens (it's not used to derivate new XP values or anything).

Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:1345]: ye I realised.

File: src/main/java/carpet/script/value/EntityValue.java [hunk 4]
```diff
@@ -1294,6 +1319,26 @@ else if (v instanceof ListValue)
...
+        });
+
+        put("score", (e, v)->{
+            if(e instanceof PlayerEntity) ((PlayerEntity) e).setScore((int) NumericValue.asNumber(v).getLong());
+        });
+
+        put("add_score",(e, v)->{
+            if(e instanceof PlayerEntity) ((PlayerEntity) e).addScore((int) NumericValue.asNumber(v).getLong());
+        });
+
```
Comment (altrisi) [inline src/main/java/carpet/script/value/EntityValue.java:1341]: Maybe make style a bit more consistent with the rest (adds spaces) (ignore that in the reviews it seems to have multiple lines):
```suggestion
        put("absorption", (e, v) -> {
            if (e instanceof PlayerEntity) ((PlayerEntity) e).setAbsorptionAmount((float) NumericValue.asNumber(v).getLong());
        });

        put("add_xp", (e, v) -> {
            if (e instanceof PlayerEntity) ((PlayerEntity) e).addExperience((int) NumericValue.asNumber(v).getLong());
        });

        put("add_xp_level", (e, v) -> {
            if (e instanceof PlayerEntity) ((PlayerEntity) e).addExperienceLevels((int) NumericValue.asNumber(v).getLong());
        });

        put("score", (e, v) -> {
            if (e instanceof PlayerEntity) ((PlayerEntity) e).setScore((int) NumericValue.asNumber(v).getLong());
        });

        put("add_score", (e, v) -> {
            if (e instanceof PlayerEntity) ((PlayerEntity) e).addScore((int) NumericValue.asNumber(v).getLong());
        });
```

File: src/main/java/carpet/script/value/EntityValue.java [hunk 5]
```diff
@@ -476,6 +476,31 @@ public Value get(String what, Value arg)
...
+        put("xp_points", (e, a)->{//todo add modify implementation to docs.
+            if(e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).experienceProgress);
+            return Value.NULL;
+        });
+
+        put("score", (e, a)->{
+            if(e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).getScore());
+            return Value.NULL;
+        });
```
Comment (altrisi) [inline src/main/java/carpet/script/value/EntityValue.java:502]: Style consistency again here. Also shouldn't the TODO be removed?
```suggestion
        put("absorption", (e, a) -> {
            if (e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).getAbsorptionAmount());
            return Value.NULL;
        });

        put("xp", (e, a) -> {
            if (e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).totalExperience);
            return Value.NULL;
        });

        put("xp_level", (e, a) -> {
            if (e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).experienceLevel);
            return Value.NULL;
        });

        put("xp_points", (e, a) -> {//todo add modify implementation to docs.
            if (e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).experienceProgress);
            return Value.NULL;
        });

        put("score", (e, a) -> {
            if (e instanceof PlayerEntity) return new NumericValue(((PlayerEntity) e).getScore());
            return Value.NULL;
        });
```

File: src/main/java/carpet/script/api/Auxiliary.java [hunk 6]
```diff
@@ -453,6 +453,30 @@ else if (!interactable && targetBlock == null)
...
+                }
+            }   else if(pVal instanceof EntityValue && ((EntityValue)pVal).getEntity() instanceof PlayerEntity){
+                p =(PlayerEntity) ((EntityValue)pVal).getEntity();
+                p.sendMessage(message,true);
+            }   else
+                throw new InternalExpressionException("'action_bar' requires a player as first argument");
+            
+            return (_c, _t) -> Value.NULL;
+        });
+
```
Comment (altrisi) [inline src/main/java/carpet/script/api/Auxiliary.java:479]: Guess what? Style again. Also in other functions there are no empty newlines between things, and IEE should also suggest a list.
```suggestion
        expression.addLazyFunction("action_bar", 2, (c, t, lv) ->
        {
            Value pVal = lv.get(0).evalValue(c);
            PlayerEntity p;
            Text message = Text.of(lv.get(1).evalValue(c).getString());
            if (pVal instanceof ListValue)
            {
                List<Value> players = ((ListValue) pVal).getItems();
                for (Value player:players)
                {
                    if (!(player instanceof EntityValue && ((EntityValue) pVal).getEntity() instanceof PlayerEntity))
                        throw new InternalExpressionException("'action_bar' only takes players as arguments");
                    p = (PlayerEntity) ((EntityValue)player).getEntity();
                    p.sendMessage(message, true);
                }
            }
            else if (pVal instanceof EntityValue && ((EntityValue) pVal).getEntity() instanceof PlayerEntity)
            {
                p = (PlayerEntity) ((EntityValue) pVal).getEntity();
                p.sendMessage(message, true);
            }
            else
                throw new InternalExpressionException("'action_bar' requires a player or a player list as first argument");
            return (_c, _t) -> Value.NULL;
        });
```

**Review by altrisi — APPROVED** [review]
Looks good. Found a few style inconsistencies with the rest of the class, if you care about those.

File: src/main/java/carpet/script/api/Auxiliary.java (hunk 6 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/api/Auxiliary.java:479]: o ye forgot.

**Comment by altrisi** [comment]
The last commit doesn't fully fix #547 since the issue also asks for querying.

File: src/main/java/carpet/script/value/EntityValue.java (hunk 4 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:1341]: ok, i dont rly care, but whatever.

File: docs/scarpet/api/Auxiliary.md [hunk 7]
```diff
@@ -191,6 +191,9 @@ produce an exception.
 
 Displays the result of the expression to the chat. Overrides default `scarpet` behaviour of sending everyting to stderr.
 
+### `action_bar(player, message)`, `action_bar(player_list,message)`
```
Comment (gnembon) [inline docs/scarpet/api/Auxiliary.md:194]: This feels redundant, especially that you already expect the player. `print` aready takes optional player - make it a third argument for print
`print(message)`, `print(player, message)`, `print(target, player, message`
in this case target could be 'chat' (default), 'title', 'subtitle', 'action_bar' etc (isn't action bar a subtitle)

This way you could also add a player array support to print.

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:194]: ok boss.

Comment (altrisi) [reply docs/scarpet/api/Auxiliary.md:194]: I personally see titles (and subtitles and actionbar) different from chat, and think that they should have a different call, since, as opposed to chat, titles can have their fade and stay times customized, feature that print can't provide (at least in its current state).

Comment (gnembon) [reply docs/scarpet/api/Auxiliary.md:194]: if that can be customized (don't know, didn't play with datapacks much before), then it makes sense to provide a separate endpoint for them, leaving print just doing chat.

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:194]: ye I can't find the method for sending titles to the player...

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:194]: Ill resend to screen.

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:194]: Illl just add it to print with a switch statement so I can expand when I find method to send a title.

Comment (altrisi) [reply docs/scarpet/api/Auxiliary.md]: The way you change title times in vanilla (don't know about datapacks) is by first running `/title times [in] [stay] [out]`, then run regular `/title`. It's quite tedious needing to run 2 commands for one action, but it's what it has.

**Comment by altrisi** [comment]
To send titles (and subtitles) you can take a look at `TitleCommand` at `net.minecraft.server.command` (yarn maps), but basically it's sending a `Title2SCPacket` to each player who will receive it. You can use the same class to send actionbars too, just get a different item from the `Action` enum (also since action bar also supports those custom times).

Something like:

```java
(ServerPlayerEntity e).networkHandler.sendPacket(new TitleS2CPacket(Title2SCPacket.Action.TITLE, text, in, stay, out));
```

Probably reusing the packet when multiple players are specified.

**Comment by altrisi** [comment]
Ok I did the title things by myself in #555.

File: docs/scarpet/api/Auxiliary.md [hunk 8]
```diff
@@ -187,9 +187,9 @@ Other value types will only be converted to tags (including NBT tags) if `force`
 extra treatment when loading them back from NBT, but using `force` true will always produce output / never 
 produce an exception.
 
-### `print(expr)`
+### `print(expr)`, `print(player/player_list, expr)`
 
-Displays the result of the expression to the chat. Overrides default `scarpet` behaviour of sending everyting to stderr.
+Displays the result of the expression to the chat. Overrides default `scarpet` behaviour of sending everyting to stderr. Can optionally define player or list of players to send the message to
```
Comment (altrisi) [inline docs/scarpet/api/Auxiliary.md:192]: The second behaviour wasn't kept when reverting actionbar stuff.

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:192]: ye I noticed, I just didn't have the time to re add that bit.

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:192]: IF u want u can do dat

File: docs/scarpet/api/Entities.md [hunk 9]
```diff
@@ -698,6 +710,21 @@ Will set entity on fire for `ticks` ticks. Set to 0 to extinguish.
...
 
+### `modify(e, 'absorption', value)`
+
+Sets the absorption value for the player. Each point is half a yellow heart.
+
+### `modify(e, 'score', value)`
+
+Sets the score of a player. Doesn't affect player xp, and is displayed upon death.
+
+### `modify(e, 'add_xp', value)`
```
Comment (gnembon) [inline docs/scarpet/api/Entities.md:721]: this exposes underlying implementation, which is not necessarily a good thing. `modify` should in general 'set' stuff unless its not practical. With modify(p, 'xp') or 'xp_level' you could move them back and forth.

**Comment by gnembon** [comment]
Suggested replacing 'add_....' to set '....' since its more flexible this way.

File: docs/scarpet/api/Entities.md (hunk 9 above)
Comment (Ghoulboy78) [reply docs/scarpet/api/Entities.md:721]: thing is, I tried that for wherever I could. In many cases I simply coulndt

Comment (Ghoulboy78) [reply docs/scarpet/api/Entities.md:721]: like there is no function for that.

Comment (Ghoulboy78) [reply docs/scarpet/api/Entities.md:721]: I got that logic, and realised it wasn't practical, except in the case of `score`, which is why I added that.

Comment (gnembon) [reply docs/scarpet/api/Entities.md:721]: Ok, I checked the XP malarchy that is in vanilla.
What makes sense to me - points are little irrelevant. We should have
for setters:
'add_xp' - nonnegative? - maybe could get it to work with negatives as well, since seems like addExperience can handle negative numbers as well, adds xp points the way vanilla `/xp add X N points` works (will change progress and level, and score)
'xp_level' - sets a level value directly, does what `/xp set X N levels`
'xp_progress' - sets directly player.experienceProgress, requires float from 0 to 1
'xp_score' - directly mess up with the score values. (using setScore)

for getters:

rename `xp_points` to `xp_progress` since it returns a float from 0 to 1

What would be helpful was to add `experience_points(level)`  (separate function in Auxiliary) that rips of getNextLevelExperience for custom input level (you will need to copy-paste it, with a reference for the future)

File: docs/scarpet/api/BlocksAndWorldAccess.md [hunk 10]
```diff
@@ -212,6 +212,15 @@ Causes a block to be harvested by a specified player entity. Honors player item
...
 that block, a block doesn't get destroyed either.
 
+### `weather(clear_ticks, rain_ticks, thunder)`,`weather(type)`
+
+Sets weather to clear for `ticks` ticks, after which it will rain for `second_ticks`
+ticks, and if `thunder` is true, it will thunder as well.
+
+If only 1 argument is specified, it must be either `'clear'`,`'rain'` or `'thunder'`, and will return true if the weather
+is in that state. NB: It can be thundering without there being a thunderstorm, there must be both thunder _and_ rain for
+there to be a thunderstorm.
```
Comment (gnembon) [inline docs/scarpet/api/BlocksAndWorldAccess.md:222]: That is a little weird syntax. 
weather() -> returns 'clear', 'rain' or 'thunder' - current weather
weather(['clear', 'rain', 'thunder']) -> returns number of ticks remaining, or '0' if not in that mode
weather(mode, tick) -> same as /weather <mode> <ticks>

Not sure what would be the use of setting the ticks or clear weather with set ticks of rain weather. In these cases the app maker can set that directly calling weather again.

**Comment by gnembon** [comment]
Suggested more intuitive use of weather and some solution for xp malarchy (its not your fault - its vanilla)

File: docs/scarpet/api/BlocksAndWorldAccess.md (hunk 10 above)
Comment (Ghoulboy78) [reply docs/scarpet/api/BlocksAndWorldAccess.md:222]: ye but the function is built in like that.

Comment (Ghoulboy78) [reply docs/scarpet/api/BlocksAndWorldAccess.md:222]: nvm ill do dat

Comment (gnembon) [reply docs/scarpet/api/BlocksAndWorldAccess.md:222]: The function is built like that but it ain't used like that. API should not be restricted by contract of internal functions but adapt them to what makes sense - /weather command does exactly that.

**Comment by Ghoulboy78** [comment]
I did the mixin thing on my own, and it worked first try!

**Comment by Ghoulboy78** [comment]
but setting thunder dont work

**Comment by Ghoulboy78** [comment]
wait I just re-read #539 and realised that it also asked for gamerules. @altrisi does the `system_settings()` func add support for those? cos I feel it would be more appropriate there...

File: src/main/java/carpet/script/api/Auxiliary.java [hunk 11]
```diff
@@ -451,7 +451,7 @@ else if (!interactable && targetBlock == null)
             {
                 ServerPlayerEntity player = EntityValue.getPlayerByValue(s.getMinecraftServer(), res);
                 if (player == null) return LazyValue.NULL;
-                s = player.getCommandSource();
+                playersVal = res;
```
Comment (gnembon) [inline src/main/java/carpet/script/api/Auxiliary.java:454]: ??

Comment (Ghoulboy78) [reply src/main/java/carpet/script/api/Auxiliary.java:454]: No, cos it could be a list of players, so I need to shift that check, just realised.

Comment (Ghoulboy78) [reply src/main/java/carpet/script/api/Auxiliary.java:454]: wait IDEA didnt tell me that u removed that var's initialization

Comment (Ghoulboy78) [reply src/main/java/carpet/script/api/Auxiliary.java:454]: ocrap that whole thing got deleted.

**Comment by Ghoulboy78** [comment]
ye can't do the xp_progress thing, but otherwise its done

**Comment by altrisi** [comment]
> 
> 
> wait I just re-read #547 and realised that it also asked for gamerules. @altrisi does the `system_settings()` func add support for those? cos I feel it would be more appropriate there...

No, `system_info()` doesn't currently have gamerules. If added there then they would be read-only.

**Comment by Ghoulboy78** [comment]
k ill do gamerules

**Comment by Ghoulboy78** [comment]
Should it go in Auxiliary or WorldAccess? im adding it in Aux for now

**Comment by Ghoulboy78** [comment]
k its gud.

**Comment by Ghoulboy78** [comment]
ah wait I havent committed

File: src/main/resources/carpet.mixins.json [hunk 12]
```diff
@@ -170,6 +170,7 @@
     "PointOfInterest_scarpetMixin",
     "ServerWorld_scarpetEventMixin",
```
Comment (gnembon) [inline src/main/resources/carpet.mixins.json:171]: Curious, why just don't add to ServerWorld_scarpetEventMixin and potentially (its not events) change its name to ServerWorld_scarpetMixin?

Comment (Ghoulboy78) [reply src/main/resources/carpet.mixins.json:171]: wait what?

Comment (Ghoulboy78) [reply src/main/resources/carpet.mixins.json:171]: didnt notice lol

File: src/main/java/carpet/script/value/EntityValue.java [hunk 13]
```diff
@@ -1319,8 +1344,24 @@ else if (v instanceof ListValue)
...
+
+        put("add_xp", (e, v) -> {
+            if (e instanceof PlayerEntity) ((PlayerEntity) e).addExperience(NumericValue.asNumber(v).getInt());
+        });
+
+        put("xp_level", (e, v) -> {
+            if (e instanceof PlayerEntity) ((PlayerEntity) e).addExperienceLevels(NumericValue.asNumber(v).getInt()-((PlayerEntity) e).experienceLevel);
+        });
+
+        put("xp_score", (e, v) -> {
```
Comment (gnembon) [inline src/main/java/carpet/script/value/EntityValue.java:1359]: What about modify(.., 'xp_progress') and that function to computes points per level that I mentioned earlier?

Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:1359]: k

File: src/main/resources/carpet.mixins.json (hunk 12 above)
Comment (Ghoulboy78) [reply src/main/resources/carpet.mixins.json:171]: literally 2 lines above, im blind

**Comment by gnembon** [comment]
can you revert the gamerules and get done with this PR? btw

**Comment by Ghoulboy78** [comment]
there is a map of gamerules, but its private, so imma add mixin cos idk about the accessor thing in #600

File: docs/scarpet/api/Auxiliary.md [hunk 14]
```diff
@@ -328,7 +329,12 @@ write_file('foo', 'shared_raw', 'five\n', 'six\n');
 
 read_file('foo', 'shared_text')     => ['one', 'two', 'three', '', 'four', '', 'five', 'six']
 </pre>
-  
+
+### `gamerule()`,`gamerule(rule)`,`gamerule(rule, value)`
+
+If run with no args, returns map of gamerules to their values.
+If run with just the rule name (lowercase, with underscores, e.g `do_mob_loot` instead of `doMobLoot`)
```
Comment (gnembon) [inline docs/scarpet/api/Auxiliary.md:336]: since these names are used in vanilla, lets stick to camelCase here as well.

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:336]: ye Ill update when I migrate to other pr

**Comment by Ghoulboy78** [comment]
k moved gamerule handling to #610  @gnembon

File: src/main/java/carpet/script/value/EntityValue.java (hunk 13 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:1368]: butidk what exactly should It do? set the player's xp based on the fill level of xp bar?

File: docs/scarpet/api/Auxiliary.md [hunk 15]
```diff
@@ -275,15 +276,13 @@ Available output types:
 ### `read_file(resource, type)`
 ### `delete_file(resource, type)`
 ### `write_file(resource, type, data, ...)`
-### `list_files(resource, type)`
 
 With the specified `resource` in the scripts folder, of a specific `type`, writes/appends `data` to it, reads its
- content, deletes the resource, or lists other files under this resource.
```
Comment (gnembon) [inline docs/scarpet/api/Auxiliary.md:281]: why did you remove my recent changes?

**Review by gnembon — CHANGES_REQUESTED** [review]

File: docs/scarpet/api/Auxiliary.md (hunk 15 above)
Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:281]: Idk, didnt do that on purpose. Ig the branch outdated, ill fix dat

Comment (Ghoulboy78) [reply docs/scarpet/api/Auxiliary.md:281]: ye ill fix others at pc cos phone screen sucks for coding

File: src/main/java/carpet/script/value/EntityValue.java (hunk 13 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/value/EntityValue.java:1368]: @gnembon can u explain, cos all I need to do is this before we can finally merge this age-old pr.
